libs/contrib/Data/SnocList/Operations.idr

```diff
@@ -0,0 +1,133 @@
+||| Operations on `SnocList`s, analogous to the `List` ones.
+||| Depending on your style of programming, these might cause
+||| ambiguities, so import with care
+module Data.SnocList.Operations
+
+import Data.SnocList
+import Data.List
+import Data.Nat
+import Syntax.PreorderReasoning
+import Syntax.PreorderReasoning.Generic
+
+%default total
+
+||| Take `n` last elements from `sx`, returning the whole list if
+||| `n` >= length `sx`.
+|||
+||| @ n  the number of elements to take
+||| @ sx the snoc-list to take the elements from
+public export
+ekat : (n : Nat) -> (sx : SnocList a) -> SnocList a
+ekat (S n) (sx :< x) = ekat n sx :< x
+ekat _ _ = [<]
+
+||| Remove `n` last elements from `xs`, returning the empty list if
+||| `n >= length xs`
+|||
+||| @ n  the number of elements to remove
+||| @ xs the list to drop the elements from
+public export
+pord : (n : Nat) -> (sx : SnocList a) -> SnocList a
+pord  0    sx        = sx
+pord (S n) [<]       = [<]
+pord (S n) (sx :< x) = pord n sx
+
+public export
+concatPordEkat : (n : Nat) -> (sx : SnocList a) ->
+  pord n sx ++ ekat n sx === sx
+concatPordEkat 0 (sx :< x) = Refl
+concatPordEkat (S n) (sx :< x) = cong (:< x) $ concatPordEkat n sx
+concatPordEkat 0 [<] = Refl
+concatPordEkat (S k) [<] = Refl
+
+
+{- We can traverse a list while retaining both sides by decomposing it
+   into a snoc-list and a cons-list
+-}
+
+||| Shift `n` elements from the beginning of `xs` to the end of `sx`,
+||| returning the same lists if `n` >= length `xs`.
+|||
+||| @ n  the number of elements to take
+||| @ sx the snoc-list to append onto
+||| @ xs the list to take the elements from
+public export
+shiftLeft : (n : Nat) -> (sx : SnocList a) -> (xs : List a) -> (SnocList a, List a)
```

Comment:
combinations of `(take, drop)` are usually called `splitAt` IIRC

Reply:
I see. How about `splitOntoLeft`/`splitOntoRight` then?